### PR TITLE
Avoid an assertion on the asyncify state changing during shutdown

### LINKED ISSUE #14902: `emscripten_force_exit` changing ASYNCIFY state?
I noticed this when modifying `tests/browser/async_mainloop.cpp`

If I add a call to `emscripten_force_exit` to the top of the `iter` function I see the following exception when running `./tests/runner browser.test_async_mainloop`:

```
[client logging: /?exception=Uncaught Error: import emscripten_force_exit was not in ASYNCIFY_IMPORTS, but changed the state / undefined ]
FAIL
```

### PR DESCRIPTION
**Pull request by kripken**
We added an asyncify state there so that we could detect async events
happening in atexits or after shutdown, both of which are disallowed.
That state is switched to when we exit the runtime, so we must allow
that change to happen without assertions on the state changing when it
was not expected to.

Fixes #14902

### CODE REVIEW

File: tests/browser/async_mainloop.cpp [hunk 1]
```diff
@@ -10,7 +10,7 @@
 extern "C" {
 
 void EMSCRIPTEN_KEEPALIVE finish(int result) {
-  REPORT_RESULT(result);
+  emscripten_force_exit(result);
```
Comment (sbc100) [inline tests/browser/async_mainloop.cpp:13]: Can you instead use `assert(result ==121)` here and then `emscripten_force_exit(0);` which should allow your to completely remove the `assert_returncode` argument to `btest_exit`.

I've been leaning towards that style as much as possible and I think it has several advantages.

File: src/library_async.js [hunk 2]
```diff
@@ -74,12 +74,17 @@ mergeInto(LibraryManager.library, {
...
                 if (Asyncify.state !== originalAsyncifyState &&
+                    // Changing the state from normal to disabled is allowed, as
+                    // that is what shutdown does.
+                    !(originalAsyncifyState === Asyncify.State.Normal &&
+                      Asyncify.state        === Asyncify.State.Disabled) &&
                     ASYNCIFY_IMPORTS.indexOf(x) < 0 &&
                     !x.startsWith('__asyncjs__') &&
+                    // invoke_* functions are allowed to change the state if we
+                    // do not ignore indirect calls.
                     !(x.startsWith('invoke_') && {{{ !ASYNCIFY_IGNORE_INDIRECT }}})) {
```
Comment (sbc100) [inline src/library_async.js:88]: This if condition is starting to get a little out of hand.  I won't block this change but if you can think of a way to refactor that would make this more readable I think that would be good.

**Review by sbc100 — APPROVED** [review]
Nice!  Thanks for looking into this one.

**Comment by kripken** [comment]
Sounds good, thanks, comments should now be addressed.

**Pull request merged**